### PR TITLE
Safely handle escape sequences

### PR DESCRIPTION
**Pull request by btmcg**
And accompanying tests

### CODE REVIEW

File: lib/scanner.l [hunk 1]
```diff
@@ -82,7 +82,7 @@ include_open ^[ \t]*@include[ \t]+\"
 <STRING>\\t       { scanctx_append_char(yyextra, '\t'); }
 <STRING>\\f       { scanctx_append_char(yyextra, '\f'); }
 <STRING>\\\\      { scanctx_append_char(yyextra, '\\'); }
-<STRING>\\\"      { scanctx_append_char(yyextra, '\"'); }
+<STRING>\\\"      { scanctx_append_string(yyextra, "\""); }
```
Comment (hyperrealm) [inline lib/scanner.l]: Why is this change necessary? It does the same thing either way.

Comment (btmcg) [reply lib/scanner.l]: The new code appends two characters: `\` and `"`. The old code appends a single character: `\"`, which is an escaped double-quote. Subtle difference, but see the test for proof.

Comment (hyperrealm) [reply lib/scanner.l]: Which compiler are you using? "\"" should be interpreted as a string of length one, containing a double quote character. For example with gcc:

int main(void) {
  char c = '\"';
  const char *s = "\"";
  printf("%c\n", c);
  printf("%s\n", s);
  return 0;
}

the output is:

"
"

If you wanted to append \" it would have to be "\\\""

Comment (hyperrealm) [reply lib/scanner.l]: Ugh... the HTML unescaped my escapes...added spaces to make it clear:

char c = " \ ' ";
const char *s = " \ " ";

Comment (btmcg) [reply lib/scanner.l]: The difference is that one is a string literal while the other is a char literal. A string literal has the implicit trailing null.

Comment (btmcg) [reply lib/scanner.l]: I misspoke before. The two characters being appended in the new code are in fact `"` and `NUL`.

Comment (hyperrealm) [reply lib/scanner.l]: Ah, I see now. The problem is in strbuf_append_char(). It needs to append a NUL byte after appending the character.

void strbuf_append_char(strbuf_t *buf, char c)
{
  strbuf_ensure_capacity(buf, 1);
  *(buf->string + buf->length) = c;
  ++(buf->length);
  *(buf->string + buf->length) = NUL;  // Add this.
}

Then you can revert the change in scanner.l

Comment (btmcg) [reply lib/scanner.l]: Ah, that works. Thanks. Interested in me updating the pull request to reflect the change?

Comment (hyperrealm) [reply lib/scanner.l]: Yes please. Also could you expand the new test to check the other escaped chars that were appended using strbuf_append_char(), which have the same problem. For example strings ending in a backslash.

Comment (btmcg) [reply lib/scanner.l]: Will do.

**Comment by btmcg** [comment]
Squashed and rebased into one clean commit.

**Comment by hyperrealm** [comment]
Thanks for doing this work!
